bus/client.go

```diff
@@ -328,8 +328,42 @@ func (c *Client) RecommendedFee() (fee types.Currency, err error) {
 
 // ContractsForSlab returns contracts that can be used to download the provided
 // slab.
-func (c *Client) ContractsForSlab(shards []object.Sector) (contracts []Contract, err error) {
-	panic("unimplemented")
+func (c *Client) ContractsForSlab(shards []object.Sector, contractSetName string) ([]Contract, error) {
+	// fetch all contracts from the set
+	contractIds, err := c.ContractSetContracts(contractSetName)
+	if err != nil {
+		return nil, err
+	}
+
+	// build two maps that indicate whether a contract is needed or usable, a
+	// contract is needed when there are shards located on its host, a contract
+	// is only usable if its in the contract set
+	isNeeded := make(map[string]struct{})
+	for _, shard := range shards {
+		isNeeded[shard.Host.String()] = struct{}{}
+	}
+	isUsable := make(map[types.FileContractID]struct{})
+	for _, contractId := range contractIds {
+		isUsable[contractId.ID] = struct{}{}
+	}
+
+	// fetch all contracts and filter them
+	contracts, err := c.Contracts()
```

Comment:
I originally wrote this temporary implementation of `ContractsForSlab` to get to a working upload/download quickly. The `Client` should be asking the `bus` rather than figuring it out on its own. However, the fact that contracts and contract sets are in different stores makes this slightly awkward.

I was thinking maybe a contract set does not need to be in its own store per se. What if the contract set relations were defined by a many-to-many between a `dbContract` and a contract set name. It could be its own entity but that's probably not even necessary if you have `contract_id <-> set_name`.

I think that would be elegant because you would then be able to ask the bus for `c.Contracts(setname)`, adding/removing contracts from the set becomes a little bit more annoying but not really.

Reply:
See https://github.com/SiaFoundation/renterd/pull/68